joint_trajectory_controller/src/joint_trajectory_controller.cpp

```diff
@@ -583,17 +585,18 @@ void JointTrajectoryController::feedback_setup_callback(
       goal_handle->get_goal()->trajectory);
 
     add_new_trajectory_msg(traj_msg);
-
-    RealtimeGoalHandlePtr rt_goal = std::make_shared<RealtimeGoalHandle>(goal_handle);
-    rt_goal->preallocated_feedback_->joint_names = joint_names_;
-    rt_active_goal_ = rt_goal;
-    rt_active_goal_->execute();
   }
 
+  // Update the active goal
+  RealtimeGoalHandlePtr rt_goal = std::make_shared<RealtimeGoalHandle>(goal_handle);
+  rt_goal->preallocated_feedback_->joint_names = joint_names_;
+  rt_goal->execute();
+  rt_active_goal_.writeFromNonRT(rt_goal);
+
   // Setup goal status checking timer
   goal_handle_timer_ = node_->create_wall_timer(
     action_monitor_period_.to_chrono<std::chrono::seconds>(),
-    std::bind(&RealtimeGoalHandle::runNonRealtime, rt_active_goal_));
+    std::bind(&RealtimeGoalHandle::runNonRealtime, rt_goal));
 }
```

Comment:
I think this needs further work, probably in a follow-up PR.

In both the old and new code, I worry that we might end up in a condition where we're overwriting the `goal_handle_timer_` while the previous `rt_active_goal_` has pending operations, and thus dropping those operations. I think we should call `goal_handle_timer_::execute_callback()` or `rt_active_goal_.readFromNonRT()->runNonRealtime()` before creating the new timer.

Similarly, I think we could probably end up in a race condition where something is resetting the `rt_active_goal_` right after this function writes the new `rt_goal`, and we could end up losing the goal handle. This would involve a little more work to solve, for example only resetting the `rt_active_goal_` if `new_data_available_ == false`. That variable is private, but something along those lines.

Would appreciate your thoughts so we can consider appropriate follow-up PRs.

Reply:
This write-up above is a perfect start for the description of a follow-up issue ;)

Reply:
@destogl beat me to it! See #166